Important bug fix, and a decent implementation of async calls.

ShafiTek opened the pull request:
- Asynchronously call Mask RCNN API and Google Vision API. After both
call succeeds, calculate the score and call /api/update. On /api/update
success, go to the post page. Check script.js for more info.
- Updated the Rate page. Results do not show up in that page anymore.
- Fixed bugs. Made `img_post` in PredictedImagePost OneToOneField and
made it as the primary key. Updated the admin page PredictedImgPost's
display function to reflect that.
- Added proper form validation with error messages using Bootstrap alerts. Check it out by making
requests without the server on or by choosing invalid images or by 
typing in invalid email addresses.
- Changed homepage background, didn't do too much to it. But made some
minor design tweaks.
- Added http 404 page. Only works in production mode. So ignore this for
now.

I making this PR mainly because I found another bug in one of our models.  In PredictedImagePost model, the primary key should be the same as ImgPost's model. This issue has now been addressed by making PredictedImagePost.img_post a one-one field.

kennylee9672 commented:
Looks good to me. I gonna fix the format in the style.css the next PR.

fstamenkovic commented on PP_APP/server/frontend/static/js/script.js:
```diff
@@ -3,126 +3,87 @@ let user_email = document.getElementById('user_email');
+let host = window.location.host;
+let pathArray = window.location.pathname.split('/');
+
+let GV_HOST = "http://127.0.0.1:8080";
```
This is a local GV host; there is a dedicated server for this, the host is 34.71.6.144. We agreed that Flask and Django should not run on the same server due to possible traffic issues, so I spent a decent amount of time configuring a VM to serve the Flask app. I think we should stick to the architecture we outlined because the infrastructure for the Flask app running on a separate VM has already been built.

All documentation is in the flask_app directory, and I have already verified that it works when writing my own ajax calls.

ShafiTek replied:
The Javascript is something that is downloaded to the client-side by the browser before it executes. Therefore, It has nothing to do with Django or Flask. I installed Flask locally and it runs on a different port than Django (which runs on 8000) for testing purposes. And also because it is better for cost reasons as each server call accumulates cost. This is perfectly in accordance with our architecture.

fstamenkovic replied:
We can host flask on on the same VM as Django, but for the record I have been hosting the flask app on Google Cloud constantly for the past month to make it available for everyone because that is what our architecture outlined. If this wasn't the plan or a part of our architecture, I should have been told when I submitted my PR's that emphasized that it's hosted on it's own VM to save money and time.

fstamenkovic commented on PP_APP/server/frontend/static/js/script.js:
```diff
@@ -3,126 +3,87 @@ let user_email = document.getElementById('user_email');
+      mrcnn_in_data.append("top_img", top_input.files[0]);
+      mrcnn_in_data.append("side_img", side_input.files[0]);
+      
+      gv_in_data_top.append("side_view", side_input.files[0]);
```
This should be: 
`gv_in_data_top.append("side_view", top_input.files[0]);`
if you want to enter the top image.

I apologize for the misnomer here; the side_view field is not the best naming practice.

ShafiTek replied:
Got it. Fixed.

fstamenkovic commented on PP_APP/server/frontend/templates/rate.html:
```diff
@@ -4,12 +4,21 @@
         <div class="row">
             <div class="col-sm">
-                <label id="top-dropzone" class="dropzone top-dropzone" style="background:linear-gradient(to bottom, rgba(22, 22, 22, 0.5) 0%, rgba(22, 22, 22, 0.8) 80%, #0000008a 100%), url({% static 'img/sample_top.jpg' %}) top left no-repeat;background-size:cover;">                    
+                <label id="top-dropzone" class="dropzone top-dropzone" style="background: url({% static 'img/sample_top.png' %}) top left no-repeat;background-size:cover;">                    
```
The white background and the white text are really difficult to distinguish. You probably edited these images yourself, maybe try giving them a darker background?

fstamenkovic commented on PP_APP/server/frontend/templates/post_view.html:
```diff
@@ -0,0 +1,59 @@
+                    <td>{{key.img_post.user_id.email}}</td>
+                </tr>
+                <tr>
+                    <th>Top</th>
```
Great page. After the beta we can actually display these images instead of providing a link.

fstamenkovic commented on PP_APP/server/frontend/templates/post_view.html:
```diff
@@ -0,0 +1,59 @@
+                    <td>{{key.packager.brand_name}}</td>
+                </tr>
+                <tr>
+                    <th>Materials</th>
```
We will also parse this after the beta to show a response in English instead of a JSON response. I can create a Github Issue for this to keep track of it.

fstamenkovic commented on PP_APP/server/frontend/templates/not_found_404.html:
```diff
@@ -0,0 +1,14 @@
+    </div>
+</section>
+
+{% endblock content %}
```
Great that we have this page. Also something to do after the beta, but we should create some functionality in case mask-rcnn or GV fail for the user to submit their own grading. Or at least let them know what happened.

fstamenkovic commented on PP_APP/server/frontend/templates/post_view.html:
```diff
@@ -0,0 +1,59 @@
+                    <td>{{key.materials}}</td>
+                </tr>
+                <tr>
+                    <th>Score</th>
```
Another thing for after the beta; tell them how we calculated the score. We can put a thorough explanation of the score in a page and put a link here so they can go and read about it. But this is great for now.

ShafiTek commented:
> * There

Yeah, no problem! 

I will merge `pp_web3_amend` as it involves less steps. However, I am not going to touch anything HTML for the time being. I think it's better for you guys to address some of the design concerns outlined here and by the client.

ShafiTek commented:
> This is terrific work. There are two minor things that need to be addressed before I'll approve this to be merged to master.
> 
> 1. The use of Flask app. We can use a separate VM that is up and running just by changing the host.
> 2. Git history; Ryland and I already implemented the ajax calls for GV yesterday and it'd be great to include those commits too. There are two options for this:
> 
> * There is a branch called `pp_web3_amend` that incorporates them without a change to the end result. I quickly created it and ran git diff on `pp_web3` and `pp_web3_amend`; they are identical in the end result. You could address the other minor feedback directly there and just merge that branch to master.
> * If you want to do it personally, all you would need to do is to rebase from branch `ajax_grading`. There'll be a conflict in two files because we changed some of the same lines, but doing `git checkout --theirs file_name` for both files will take your version. Just make sure everything is right before you force push. Feel free to drop me a note about this.
> 
> Option number one would save you a few minutes; you can address other feedback on that branch directly.
> 
> Awesome work man and a big thank you!



> Code is good, but like Filip, I would like if the other commits could be included as well. Nice work

Done.
